File: skeletons/dotnet-angular-modular-monolith-docker-compose/src/MyApp.Contracts/Exceptions.cs
namespace MyApp.Contracts;

/// <summary>Typed application errors mapped to Problem Details (adrs/dotnet/rfc7807-errors.md).</summary>
public abstract class AppException(string detail) : Exception(detail)
{
    public abstract int StatusCode { get; }
    public abstract string Title { get; }
}

public sealed class NotFoundException(string detail) : AppException(detail)
{
    public override int StatusCode => StatusCodes.Status404NotFound;
    public override string Title => "Not Found";
}

public sealed class ConflictException(string detail) : AppException(detail)
{
    public override int StatusCode => StatusCodes.Status409Conflict;
    public override string Title => "Conflict";
}

public sealed class BadRequestException(string detail) : AppException(detail)
{
    public override int StatusCode => StatusCodes.Status400BadRequest;
    public override string Title => "Bad Request";
}

public sealed class UnauthorizedException(string detail) : AppException(detail)
{
    public override int StatusCode => StatusCodes.Status401Unauthorized;
    public override string Title => "Unauthorized";
}

public sealed class ForbiddenException(string detail) : AppException(detail)
{
    public override int StatusCode => StatusCodes.Status403Forbidden;
    public override string Title => "Forbidden";
}

internal static class StatusCodes
{
    public const int Status400BadRequest = 400;
    public const int Status401Unauthorized = 401;
    public const int Status403Forbidden = 403;
    public const int Status404NotFound = 404;
    public const int Status409Conflict = 409;
}

File: skeletons/python-react-modular-monolith-docker-compose/src/app/core/exceptions.py
import logging

from fastapi import FastAPI, Request, status
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from starlette.exceptions import HTTPException as StarletteHTTPException

logger = logging.getLogger(__name__)

PROBLEM_MEDIA_TYPE = "application/problem+json"


class AppError(Exception):
    """Base typed application error (adrs/python/rfc7807-errors.md)."""

    status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
    title = "Internal Server Error"
    headers: dict[str, str] | None = None

    def __init__(self, detail: str) -> None:
        super().__init__(detail)
        self.detail = detail


class NotFoundError(AppError):
    status_code = status.HTTP_404_NOT_FOUND
    title = "Not Found"


class ConflictError(AppError):
    status_code = status.HTTP_409_CONFLICT
    title = "Conflict"


class BadRequestError(AppError):
    status_code = status.HTTP_400_BAD_REQUEST
    title = "Bad Request"


class UnauthorizedError(AppError):
    status_code = status.HTTP_401_UNAUTHORIZED
    title = "Unauthorized"
    headers = {"WWW-Authenticate": "Bearer"}


class ForbiddenError(AppError):
    status_code = status.HTTP_403_FORBIDDEN
    title = "Forbidden"


def problem(
    status_code: int,
    title: str,
    detail: str,
    headers: dict[str, str] | None = None,
    **extensions: object,
) -> JSONResponse:
    body: dict[str, object] = {"title": title, "status": status_code, "detail": detail}
    body.update(extensions)
    return JSONResponse(
        body, status_code=status_code, media_type=PROBLEM_MEDIA_TYPE, headers=headers
    )


def register_exception_handlers(app: FastAPI) -> None:
    @app.exception_handler(AppError)
    async def app_error_handler(request: Request, exc: AppError) -> JSONResponse:
        return problem(exc.status_code, exc.title, exc.detail, headers=exc.headers)

    @app.exception_handler(RequestValidationError)
    async def validation_handler(request: Request, exc: RequestValidationError) -> JSONResponse:
        errors = [
            {"field": ".".join(str(part) for part in e["loc"][1:]), "message": e["msg"]}
            for e in exc.errors()
        ]
        return problem(
            422,
            "Validation Failed",
            "The request body failed validation.",
            errors=errors,
        )

    @app.exception_handler(StarletteHTTPException)
    async def http_exception_handler(request: Request, exc: StarletteHTTPException) -> JSONResponse:
        return problem(exc.status_code, "Error", str(exc.detail))

    @app.exception_handler(Exception)
    async def unhandled_handler(request: Request, exc: Exception) -> JSONResponse:
        logger.exception("unhandled error on %s %s", request.method, request.url.path)
        return problem(
            status.HTTP_500_INTERNAL_SERVER_ERROR,
            "Internal Server Error",
            "An unexpected error occurred.",
        )
